maven-plugin/src/main/java/io/confluent/kafka/schemaregistry/maven/derive/schema/json/DeriveJsonSchema.java

```diff
@@ -0,0 +1,130 @@
+    // Generate Schema for Primitive type
+    String jsonInferredType = field.getClass().getName();
+    if (classToDataType.containsKey(jsonInferredType)) {
+      String schemaString = String.format("{\"type\" : \"%s\"}",
+          classToDataType.get(jsonInferredType));
+      return Optional.of(mapper.readValue(schemaString, ObjectNode.class));
+    }
+    return Optional.empty();
+  }
+
+  private static ArrayList<ObjectNode> getSchemaOfAllElements(List<JsonNode> messages, String name)
+      throws JsonProcessingException {
+    // Get schema of each message based on type
+    ArrayList<ObjectNode> schemaList = new ArrayList<>();
+    for (JsonNode message : messages) {
+      schemaList.add(getSchemaOfElement(message, name));
+    }
+    return schemaList;
+  }
+
+  private static ObjectNode getSchemaOfElement(JsonNode message, String name)
+      throws JsonProcessingException {
+    // Get schema of message based on type
+    Optional<ObjectNode> primitiveSchema = getPrimitiveSchema(message);
+    if (primitiveSchema.isPresent()) {
+      return primitiveSchema.get();
+    } else if (message instanceof ArrayNode) {
+      return getSchemaForArray(DeriveSchema.getListFromArray((ArrayNode) message), name);
+    } else {
+      ObjectNode objectNode = mapper.valueToTree(message);
+      return getSchemaForRecord(objectNode);
+    }
+  }
+
+  public static ObjectNode getSchemaForArray(List<JsonNode> messages, String name)
+      throws JsonProcessingException {
+    // Generate Schema for Array type
+    ObjectNode schema = mapper.createObjectNode();
+    schema.put("type", "array");
+    ArrayList<ObjectNode> schemaList = getSchemaOfAllElements(messages, name);
+    ObjectNode items = mergeArrays(schemaList, false);
+    schema.set("items", items.get("items"));
+    return schema;
+  }
+
+  public static ObjectNode getSchemaForRecord(ObjectNode message)
+      throws JsonProcessingException {
+    // Generate Schema for Record type
+    ObjectNode schema = mapper.createObjectNode();
+    schema.put("type", "object");
+    schema.set("properties", mapper.createObjectNode());
+
+    // Loop over each field, get type of each field and insert into schema
+    for (String fieldName : DeriveSchema.getSortedKeys(message)) {
+      JsonNode field = message.get(fieldName);
+      ObjectNode fields = (ObjectNode) schema.get("properties");
+      fields.set(fieldName, getSchemaOfElement(field, fieldName));
+    }
+    return schema;
+  }
+
+  public static ObjectNode getSchemaForMultipleMessages(List<String> messages)
+      throws JsonProcessingException {
+    /*
+     Get schema for multiple messages. Exactly one schema is returned
+     Treated same as array of records, the items derived is returned
+     */
+    List<JsonNode> messageObjects = new ArrayList<>();
+    for (String message : messages) {
+      messageObjects.add(mapper.readTree(message));
+    }
+    ObjectNode schema = (ObjectNode) getSchemaForArray(
```

Comment:
Unnecessary casting could be avoided.
```suggestion
    JsonNode schema = getSchemaForArray(
```